TLS 1.3: Refine and test client early data status

ronald-cron-arm opened the pull request:
## Description
First PR of two addressing #6340.

## PR checklist
- [x] **changelog** not required, will add one for TLS 1.3 early data feature as a whole
- [x] **backport** not required, new feature
- [x] **tests** provided

mpg commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3787,3 +3787,146 @@ exit:
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_SSL_EARLY_DATA:MBEDTLS_SSL_CLI_C:MBEDTLS_SSL_SRV_C:MBEDTLS_TEST_AT_LEAST_ONE_TLS1_3_CIPHERSUITE:MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED:MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED:MBEDTLS_MD_CAN_SHA256:MBEDTLS_ECP_HAVE_SECP256R1:MBEDTLS_ECP_HAVE_SECP384R1:MBEDTLS_PK_CAN_ECDSA_VERIFY:MBEDTLS_SSL_SESSION_TICKETS */
+void tls13_cli_early_data_status()
```
Note: this is good, but I think at some point we also want tests for the cases where early data won't work (not sent, rejected/hrr, rejected/no extension). Could probably be done by adding a "scenario" argument to this function?

Edit: done in the next few commits :)

mpg commented on tests/suites/test_suite_ssl.function:
```diff
@@ -14,8 +14,9 @@
-#define TEST_EARLY_DATA_DEPROTECT_AND_DISCARD 1
-#define TEST_EARLY_DATA_DISCARD_AFTER_HRR 2
+#define TEST_EARLY_DATA_NO_INDICATION_SENT 1
+#define TEST_EARLY_DATA_DEPROTECT_AND_DISCARD 2
```
Note: this name an the next one are a bit server-centric. Perhaps "not accepted" and "rejected hrr" would be more neutral names?

Edit: done in the next two commits :)

mpg reviewed (commented):
Looks pretty good to me! I only have minor suggestions for improvements (mostly comments, naming, etc.) but everything looks correct and well-tested to me.

tom-cosgrove-arm commented on include/mbedtls/ssl.h:
```diff
@@ -5106,9 +5106,9 @@ int mbedtls_ssl_close_notify(mbedtls_ssl_context *ssl);
-#define MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_SENT  0
-#define MBEDTLS_SSL_EARLY_DATA_STATUS_ACCEPTED  1
-#define MBEDTLS_SSL_EARLY_DATA_STATUS_REJECTED  2
+#define MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_SENT  1
```
(pre-existing, minor) feels like this should be an enum

ronald-cron-arm replied:
see https://github.com/Mbed-TLS/mbedtls/pull/8755#issuecomment-1946215759

tom-cosgrove-arm commented on library/ssl_misc.h:
```diff
@@ -731,14 +731,23 @@ struct mbedtls_ssl_handshake_params {
+     *  one CCS per handshake without having to complicate the handshake state
+     *  transitions.
+     */
+    uint8_t ccs_count;
```
I agree with Manuel, but minor, can be done in a follow-up PR

ronald-cron-arm replied:
I agree as well, done in #8760 commit "tls13: Use a flag not a counter for CCS and HRR handling".

tom-cosgrove-arm commented on library/ssl_misc.h:
```diff
@@ -731,14 +731,23 @@ struct mbedtls_ssl_handshake_params {
+
+#if defined(MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE)
+    /**
+     *  Number of dummy change_cipher_spec (CCS) record sent. Used to send only
```
(minor, fix in follow-up) should be plural `records sent`. Also, there seem to be two spaces after the `*` on each line :)

ronald-cron-arm replied:
comment changed in #8760

tom-cosgrove-arm commented on library/ssl_misc.h:
```diff
@@ -2136,6 +2145,38 @@ int mbedtls_ssl_tls13_write_early_data_ext(mbedtls_ssl_context *ssl,
+ * ClientHello, the server has accepted them and the client has received the
+ * server Finished message. It cannot send early data to the server anymore.
+ */
+#define MBEDTLS_SSL_EARLY_DATA_STATUS_SERVER_FINISHED_RECEIVED 6
```
(for follow-up) If these are all possible values of `early_data_status` i.e. return values from `mbedtls_ssl_get_early_data_status()` could they be defined together, ideally in an enum?

ronald-cron-arm replied:
`mbedtls_ssl_get_early_data_status()` returns only NOT_SENT or ACCEPTED or REJECTED. The other values are for internal use only. See also https://github.com/Mbed-TLS/mbedtls/pull/8755#issuecomment-1946215759.

tom-cosgrove-arm commented on library/ssl_misc.h:
```diff
@@ -2145,6 +2145,38 @@ int mbedtls_ssl_tls13_write_early_data_ext(mbedtls_ssl_context *ssl,
                                            const unsigned char *end,
                                            size_t *out_len);
+
+#if defined(MBEDTLS_SSL_CLI_C)
```
(minor) I'd prefer to see them defined in one place, even if values are defined that aren't used under some circumstances - they are well-enough namespaced

tom-cosgrove-arm commented on tests/include/test/ssl_helpers.h:
```diff
@@ -599,6 +600,17 @@ int mbedtls_test_ticket_parse(void *p_ticket, mbedtls_ssl_session *session,
+    mbedtls_test_handshake_test_options *client_options,
+    mbedtls_test_handshake_test_options *server_options,
+    mbedtls_ssl_session *session);
+#endif /* MBEDTLS_SSL_CLI_C && MBEDTLS_SSL_SRV_C &&
```
(minor) we don't really need the list of conditions in the `#endif` since it's so close to the `#if` - it clutters the text, and if the condition needs changing in the future it's an unnecessary thing to have to keep in sync

ronald-cron-arm replied:
Done in #8760, commit "tls13: Fix/Improve comments"

tom-cosgrove-arm commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3834,3 +3788,316 @@ exit:
+                            TEST_EQUAL(client_ep.ssl.early_data_status,
+                                       MBEDTLS_SSL_EARLY_DATA_STATUS_REJECTED);
+                        }
+                        break;
```
I wonder if we should have `default: TEST_FAIL("Missing scenario");` in these in case we extend the set of scenarios and forget to update one of the `switch`es in the future?

ronald-cron-arm replied:
Yes that should reduce the risk of missing something when adding a scenario.

tom-cosgrove-arm reviewed (approved):
LGTM - I only have minor comments, and given @mpg's review is similar, and these commits are in another PR that is under review, I will add this one to the merge queue, for the comments to be addressed in a follow-up PR

tom-cosgrove-arm commented:
Thank you @mpg!

ronald-cron-arm commented on library/ssl_tls13_client.c:
```diff
@@ -3059,18 +3057,25 @@ int mbedtls_ssl_tls13_handshake_client_step(mbedtls_ssl_context *ssl)
-            if (ret == 0) {
-                mbedtls_ssl_handshake_set_state(ssl, MBEDTLS_SSL_CLIENT_HELLO);
+            ret = 0;
+            if (ssl->handshake->ccs_count == 0) {
```
Done in #8760 commit "tls13: client: Improve CCS handling"

ronald-cron-arm commented on library/ssl_tls13_client.c:
```diff
@@ -1180,26 +1180,21 @@ int mbedtls_ssl_tls13_write_client_hello_exts(mbedtls_ssl_context *ssl,
-        ssl_tls13_early_data_has_valid_ticket(ssl) &&
-        ssl->conf->early_data_enabled == MBEDTLS_SSL_EARLY_DATA_ENABLED &&
-        ssl->handshake->hello_retry_request_count == 0) {
+    if (ssl->handshake->hello_retry_request_count == 0) {
```
I agree, done in https://github.com/Mbed-TLS/mbedtls/pull/8760 commit "tls13: Use a flag not a counter for CCS and HRR handling".

ronald-cron-arm commented on library/ssl_tls13_client.c:
```diff
@@ -1180,26 +1180,21 @@ int mbedtls_ssl_tls13_write_client_hello_exts(mbedtls_ssl_context *ssl,
-        ssl->conf->early_data_enabled == MBEDTLS_SSL_EARLY_DATA_ENABLED &&
-        ssl->handshake->hello_retry_request_count == 0) {
+    if (ssl->handshake->hello_retry_request_count == 0) {
+        if (mbedtls_ssl_conf_tls13_is_some_psk_enabled(ssl) &&
```
I agree as well, done in https://github.com/Mbed-TLS/mbedtls/pull/8760 commit "tls13: client: Add comment about early data in 2nd ClientHello".

ronald-cron-arm commented on library/ssl_misc.h:
```diff
@@ -2145,6 +2145,38 @@ int mbedtls_ssl_tls13_write_early_data_ext(mbedtls_ssl_context *ssl,
+ * The client has not sent the first ClientHello yet, it is unknown if the
+ * client will send an early data indication extension or not.
+ */
+#define MBEDTLS_SSL_EARLY_DATA_STATUS_UNKNOWN 0
```
Fixed in #8760, commit "tls13: Fix/Improve comments"

ronald-cron-arm commented:
@mpg and @tom-cosgrove-arm thanks for the review.

I have addressed your comment in five additional commit in #8760 (thanks for its rebase). Regarding the definition of MBEDTLS_SSL_EARLY_DATA_STATUS_xyz (various comments about it) I've tried to improve things (last commit). MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_SENT/ACCEPTED/REJECTED documentation is still in the documentation of `mbedtls_ssl_get_early_data_status()` as they are the possible values (apart from errors) returned by this function. MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_SENT/ACCEPTED/REJECTED are now defined in the mbedtls_ssl_early_data_status enum along with the other values that are for internal use only. The enum itself is also for internal use only.
